### ISSUE TITLE
Empty string object is “_nil_”

### ISSUE THREAD

**Issue by BlakeMScurr** [dgraph-io/dgraph#1627]
Why do empty string objects actually get added as “\_nil\_“? https://github.com/dgraph-io/dgraph/blob/33154f46abd86c18487ab50bb5991b2d3228cb3e/gql/mutation.go#L167

Can this be changed so that we empty strings rather than strange `"_nil_"`results in queries like this?
```
 {
    me(func: eq(validEdge, "v")) {
      nilEdge
    }
}
```

which returns:

```
{
  "data": {
    "me": [
      {
        "nilEdge": "_nil_"
      }
    ]
  }
}
```
or this query:
```
 {
    me(func: eq(validEdge, "v")) {
      expand(_all_)
    }
}
```

which returns:
```
{
  "data": {
    "me": [
      {
        "nilEdge": "_nil_",
        "validEdge": "v"
      }
    ]
  }
}
```

given this schema:

```
mutation {
  schema {
    validEdge: string @index(exact) .
  }
}
```

and this data:

```
mutation {
    set {
      _:tester <nilEdge> "" .
      _:tester <validEdge> "v" .
    }
}
```

**Comment by peterstace** [comment]
Relates to #1484 , although looks like a separate problem.

**Comment by peterstace** [comment]
Some minimum working examples:

Shows up when the predicate is expanded.
```
mutation{
  schema{
    name: string @index(exact) .
  }
  set{
    _:a <name> "A" .
    _:a <msg> "" .
  }
}
```
```
{
  q(func: eq(name, "A")) {
    expand(_all_)
  }
}
=>
{
  "data": {
    "q": [
      {
        "msg": "_nil_",
        "name": "A"
      }
    ]
  }
}
```

Doesn't work for search (because it's stored internally as `"_nil_"` instead of `""`). Although we get an error rather than giving back the wrong result:
```
mutation{
  schema{
    name: string @index(exact) .
  }
  set{
    _:b <name> "" .
    _:b <msg> "hi" .
  }
}
```
```
{
  q(func: eq(name, "")) {
    expand(_all_)
  }
}
=>
Empty argument received
```

**Comment by peterstace** [comment]
I've put a PR up that fixes this #1646 

It just checks for `"_nil_"` during output, and makes sure to output `""` instead.

I don't think this is the ideal solution (we should just store empty strings as empty strings internally). However, it's the most pragmatic until we can tackle #1484 (probably better to wait for client to settle down so we can handle deletions differently).

**Comment by peterstace** [comment]
Cherry picked into master at cdfd6dfd2021322569d5a482e1e1e153e2cde1dc
